[quant][pt2e] Add transform_for_annotation method in Quantizer

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113115

Summary:
Adding the method so that people can do some transformations before annotation to make the graph easier to annotate

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_transform_for_annotation

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D51141080](https://our.internmc.facebook.com/intern/diff/D51141080)

jerryzh168 committed:
[quant][pt2e] Add transform_for_annotation method in Quantizer

Summary:
Adding the method so that people can do some transformations before annotation to make the graph easier to annotate

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_transform_for_annotation

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/quantizer/quantizer.py:
```diff
@@ -157,6 +157,13 @@ class QuantizationAnnotation:
 
 class Quantizer(ABC):
+
+    # transform the model before annotation, to make sure we can annotate the model properly
+    # for example, we can decompose some compound operators here, or change scalar argument to Tensor
+    # so we can annotate the argument
```
```suggestion
     # Allows for user defined transforms to run before annotating the graph. This allows quantizer to allow quantizing part of the model
     # that are otherwise not quantizable. For example quantizer can a) decompose a compound operator like scaled dot product attention, into bmm and softmax if quantizer knows how to quantize bmm/softmax but not sdpa
     # or b) transform scalars to tensor to allow quantizing scalares. 
```

kimishpatel commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -341,6 +342,11 @@ def set_module_name(
         return self
 
+    def transform_for_annotation(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+        """Transforms scalar values to tensor attributes
```
can you also add a test to xnnpack quantizer?

jerryzh168 replied:
yeah, I have this in my local PR, will update

mcr229 replied:
for the test we should also include a test where the model performs add/mul against constant sys.float_max

jerryzh168 replied:
@mcr229 is that a different issue? I think we can have a separate PR for that

mcr229 replied:
sure but i think we should add a test for that pass, with edge cases like float.sysmax/min constants

kimishpatel reviewed (approved):
Left a comment for adding test to xnnpack quantizer but feel free to defer it. and left a update to the comment about transform_for_annotation

jerryzh168 committed:
Update on "[quant][pt2e] Add transform_for_annotation method in Quantizer"


Summary:
Adding the method so that people can do some transformations before annotation to make the graph easier to annotate

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_transform_for_annotation

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/quantize_pt2e.py:
```diff
@@ -103,6 +103,7 @@ def calibrate(model, data_loader):
     # to be quantized before fusion
     # TODO: (maybe) rewrite this with subgraph_rewriter
     _fuse_conv_bn_(model)
+    quantizer.transform_for_annotation(model)
```
Should we use the same method for  QAT fusion as well? Or is it intended for something else

jerryzh168 replied:
this is not for QAT fusion, plan for QAT transforms is still to put these in the flow for now

digantdesai commented on torch/ao/quantization/quantizer/quantizer.py:
```diff
@@ -157,6 +157,17 @@ class QuantizationAnnotation:
+    # For example quantizer can
+    # a) decompose a compound operator like scaled dot product attention,
+    # into bmm and softmax if quantizer knows how to quantize bmm/softmax but not sdpa
+    # or b) transform scalars to tensor to allow quantizing scalares.
```
```suggestion
    # or b) transform scalars to tensor to allow quantizing scalars.
```

digantdesai commented on torch/ao/quantization/quantizer/quantizer.py:
```diff
@@ -157,6 +157,17 @@ class QuantizationAnnotation:
+    # a) decompose a compound operator like scaled dot product attention,
+    # into bmm and softmax if quantizer knows how to quantize bmm/softmax but not sdpa
+    # or b) transform scalars to tensor to allow quantizing scalares.
+    def transform_for_annotation(
```
Since its a public method and not part of the `annotate`, one implication is that this is optional. Is it really optional?

jerryzh168 replied:
it is, we have the default implementation that returns model directly. you can see this is not an abstractmethod that need to implemented by all quantizers.

I can add some docs to clarify.

digantdesai replied:
I understand that for a custom Quantizer it is optional to extend it. That is not what I am talking about though.

I was referring to the call to this fn from `prepare_pt2d`, which is not optional. I didn't understand the rationale for having additional call vs. something annotate would do given it has to be before annotate and not optional.

I guess from function semantics pov it might make more sense to keep it seperate from annotate.

jerryzh168 replied:
yeah this is to keep the semantics for annotate clean, annotation should not mutate the graph, I'm also planning to change the return type for annotate to None as well in the future

jerryzh168 committed:
Update on "[quant][pt2e] Add transform_for_annotation method in Quantizer"


Summary:
Adding the method so that people can do some transformations before annotation to make the graph easier to annotate

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_transform_for_annotation

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51141080).

jerryzh168 committed:
Update on "[quant][pt2e] Add transform_for_annotation method in Quantizer"


Summary:
Adding the method so that people can do some transformations before annotation to make the graph easier to annotate

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_transform_for_annotation

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D51141080](https://our.internmc.facebook.com/intern/diff/D51141080)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51141080).

jerryzh168 commented:
@pytorchbot merge

PaliC commented:
@pytorchbot revert -m "breaking executorch tests" -c ghfirst
